airflow/providers/cncf/kubernetes/utils/pod_launcher.py

```diff
@@ -117,79 +130,104 @@ def delete_pod(self, pod: V1Pod) -> None:
-                    )
-                    raise AirflowException(msg)
-                time.sleep(1)
-
-    def monitor_pod(self, pod: V1Pod, get_logs: bool) -> Tuple[State, V1Pod, Optional[str]]:
+        while True:
+            remote_pod = self.read_pod(pod)
+            if remote_pod.status.phase != PodStatus.PENDING:
+                break
+            self.log.warning("Pod not yet started: %s", pod.metadata.name)
+            delta = dt.now() - curr_time
+            if delta.total_seconds() >= startup_timeout:
+                msg = (
+                    f"Pod took longer than {startup_timeout} seconds to start. "
+                    "Check the pod events in kubernetes to determine why."
+                )
+                raise PodLaunchFailedException(msg)
+            time.sleep(1)
+
+    def follow_container_logs(self, pod: V1Pod, container_name: str):
+        """
+        Follows the logs of container and streams to airflow logging.
+        Returns when container exits.
         """
-        Monitors a pod and returns the final state, pod and xcom result
+        container_stopped = False
+        read_logs_since_sec = None
+        last_log_time = None
+
+        # `read_pod_logs` follows the logs so we shouldn't necessarily _need_ to loop
+        # but in a long-running process we might lose connectivity and this way we
+        # can resume following the logs
+        while True:
+            try:
+                logs = self.read_pod_logs(
+                    pod=pod,
+                    container_name=container_name,
+                    timestamps=True,
+                    since_seconds=read_logs_since_sec,
+                )
+                for line in logs:  # type: bytes
+                    timestamp, message = self.parse_log_line(line.decode('utf-8'))
+                    self.log.info(message)
+                    if timestamp:
+                        last_log_time = timestamp
+            except BaseHTTPError:  # Catches errors like ProtocolError(TimeoutError).
+                self.log.warning(
+                    'Failed to read logs for pod %s',
+                    pod.metadata.name,
+                    exc_info=True,
+                )
+
+            if container_stopped is True:
+                break
+
+            if last_log_time:
+                delta = pendulum.now() - last_log_time
+                read_logs_since_sec = math.ceil(delta.total_seconds())
+
+            time.sleep(1)
+
+            if self.container_is_running(pod, container_name=container_name):
+                self.log.info('Container %s is running', pod.metadata.name)
+                self.log.warning('Pod %s log read interrupted', pod.metadata.name)
+            else:
+                container_stopped = True  # fetch logs once more and exit
+
+    def await_container(self, pod: V1Pod, container_name: str) -> None:
+        while not self.container_is_running(pod=pod, container_name=container_name):
+            time.sleep(1)
```

Comment:
Do we really need a separate method here? Asking coz it is just used in a single place in the code (excluding tests)

Reply:
If we want to keep it probably rename it to `await_container_completion` ?

Reply:
> If we want to keep it probably rename it to await_container_completion ?

sure

> Do we really need a separate method here? 

personally, even though this method is very  simple, i would prefer to keep `execute` higher level.  going through and looping is to me too low-level for the `execute` method, at least in the case of this operator which already has a lot going on.  ideally execute should be easy to understand just by reading method names.